## U.S. Medical Insurance Costs

1. Import pandas as pd


In [4]:
import pandas as pd

2. Read csv file into pandas

In [6]:
insurance_data = pd.read_csv('insurance.csv')
insurance_data.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

3. Find out the typs of values in the dataset

In [8]:
insurance_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


From this query I can see that there are no missing values in the 1338 records within the dataset. The values fo sex, smoker, and region could be made in to catergories as there is a set number of each of these values.


4. Find the average age of the individuals in the dataset

In [11]:
# ages = insurance_data['age'].sum() # this line totals the values in the column 'age'

def find_average(dataset, column):
    length = len(dataset)
    total_column = insurance_data[column].sum()
    return round((total_column / length), 2)
    
average_age = find_average(insurance_data, 'age')      # round((ages / len(insurance_data)), 2)

print(f"the average age of people in the data set is {average_age} years.")

the average age of people in the data set is 39.21 years.


5. This seems representative of an adult population. Next to investigate if there is a roughly even slit between male and female customer (which is required to be representative of an adult population.

In [13]:
num_males = len(insurance_data[insurance_data['sex'] == 'male'])
num_females = len(insurance_data[insurance_data['sex'] == 'female'])

print(f"There are {num_males} males, and {num_females} females in the data set")


There are 676 males, and 662 females in the data set


The split between males and females is fairly even so I am happy this data set is representative in thisaspect of a typical adult population.

6. Find min and max values for 'age' column.

In [16]:
def min_max(dataset):
    cols = ['age']
    min_age = dataset['age'].min()
    max_age = dataset['age'].max()
    return min_age, max_age

min_age, max_age = min_max(insurance_data)
print(f"min age: {min_age} \nmax_age: {max_age}")

min age: 18 
max_age: 64


That looks better; we do not appear to have any outliers.

7. Now to check the median average to check that it correlates with the mean

In [19]:
insurance_data.describe() # Perhaps I should have done this first. 

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

The results from the method above indicates that we have a nominal data set as the mean and the median are virtually identical and the 25% and 75% quartile are roughly in line with the standard deviation.

8. And now I wish to understand the spread of the number of children amongst the data.

In [22]:
aggfuncs = ['mean', 'median', 'max', 'min']
grouped_by_children = insurance_data.groupby('children')[['age', 'bmi']].agg(aggfuncs)
grouped_by_children.head()

age                       bmi                        
               mean median max min       mean   median    max     min
children                                                             
0         38.444251   36.0  64  18  30.550296  30.3025  53.13  15.960
1         39.453704   40.0  64  18  30.623148  30.0150  52.58  16.815
2         39.445833   40.0  64  18  30.978104  31.6400  48.07  16.815
3         41.566879   41.0  64  18  30.684331  30.4950  46.53  18.905
4         39.000000   39.0  61  18  31.391200  29.6400  41.47  24.415

Looking at at the above figures it would appear that there is a nominal spread over the entire dataset, so I am now happy that this dataset is representative of an adult population, as the mean and medians for age and bmi appear to be fairly consistant regardless of number of children, although the figures for max and min bmi do have a noticeable differance between those clients that have 4 children and less than 4 children; however the mean and median for 4 children does indicate that these fall within the nominal ranges.

9. Look at the different costs between smokers vs. non-smokers. I am going to look at the average cost for a smoker and the same for a non-smoker to see how smoking affects the insrance costs for an individual. First create datasets that contain smokers and assign to var smokers. do the same for non-smokers assigning them to variable non_smokers.

In [25]:
smoker_charges_data = insurance_data[['smoker', 'charges']]
smokers = smoker_charges_data[smoker_charges_data['smoker'] == 'yes']
non_smokers = smoker_charges_data[smoker_charges_data['smoker'] == 'no']
smokers


smoker      charges
0       yes  16884.92400
11      yes  27808.72510
14      yes  39611.75770
19      yes  36837.46700
23      yes  37701.87680
...     ...          ...
1313    yes  36397.57600
1314    yes  18765.87545
1321    yes  28101.33305
1323    yes  43896.37630
1337    yes  29141.36030

[274 rows x 2 columns]

10. calculate the mean insurance costs for smokers and non-smokers, using the find_average func


In [27]:
smokers_avg_insurance_cost = find_average(smokers, 'charges')
non_smokers_avg_insurance_cost = find_average(non_smokers, 'charges')
print(f"A smoker's average insurance cost is: {smokers_avg_insurance_cost}\nA non-smoker's average insurance cost is: {non_smokers_avg_insurance_cost}")

A smoker's average insurance cost is: 64802.28
A non-smoker's average insurance cost is: 16687.81


11. It would apper that being a smoker significantly increases insurance costs from this query, however does the average age of a smoker match that of the whole dataset, as this could be another factor that significantly increases the insurance costs. I will now explore the average age of a smoker:

In [29]:
# First I need to filter smokers by age
smokers_non_smokers_by_age = insurance_data[['smoker', 'age']]
smokers_by_age = smokers_non_smokers_by_age[smokers_non_smokers_by_age['smoker'] == 'yes'] 
smokers_by_age.head()

smoker  age
0     yes   19
11    yes   62
14    yes   27
19    yes   30
23    yes   34

12. Now I will look at the describe() data from smokers_by_age

In [31]:
smokers_by_age.describe()

age
count  274.000000
mean    38.514599
std     13.923186
min     18.000000
25%     27.000000
50%     38.000000
75%     49.000000
max     64.000000

13. The mean and the median (50%) indicate that the average age of smokers in the data set is actually lower than the average cost f all customers in the dataset. this would seem to indicate that the the high insurance cost in 7 above. also the min age of a smoker in the data set is 18 and the eldest is 64, and the stnd deviation is very close to the stnd deviation in the entire dataset: this indicates this dataset is roughly in line with the overall dataset. These factors together would seem to indicate that smoking significantly raises the insurance costs of an individual. However there could be another external influence in the data to do with region, so I am now going to look at the average charges for each region.

In [33]:
aggfuncs = ['mean', 'median', 'std']
insurance_data.groupby(['region', 'smoker'])['charges'].agg(aggfuncs)

mean        median           std
region    smoker                                          
northeast no       9165.531672   8342.908750   6157.196439
          yes     29673.536473  28101.333050  11589.559770
northwest no       8556.463715   7256.723100   6128.553819
          yes     30192.003182  27488.996475  11413.827702
southeast no       8032.216309   6652.528800   6137.329758
          yes     34844.996824  37484.449300  11324.766536
southwest no       8019.284513   7348.142000   5489.514421
          yes     32269.063494  35165.256500  11270.279593

Having organised the data by region and smokers/ non-smokers, and looking at the mean and median charges, it would appear that smokers are consistantly charged significantly more across all regions within the dataset. the other factor that could contribute is BMI, so I will now investigate these.

14. First I will look at the mean, median and std deviation of BMIs in the data set

In [36]:
charges_by_bmi = insurance_data[['bmi', 'charges', 'smoker']]
charges_by_bmi.describe()

bmi       charges
count  1338.000000   1338.000000
mean     30.663397  13270.422265
std       6.098187  12110.011237
min      15.960000   1121.873900
25%      26.296250   4740.287150
50%      30.400000   9382.033000
75%      34.693750  16639.912515
max      53.130000  63770.428010

15. From the above query I can determine that the distribution of BMIs ranges from 15.96 through to 53.13 and that this also appears to be a significant factor in increasing overall insurance charges. I will now break the BMIs down into ranges of 5 (15-20, 20-25, 25-30, etc) as this will allow me to compare these ranges for a comparison between non-smokers and smokers.

In [38]:
def group_bmi_ranges(dataset):
    for i in range(len(dataset)):
        row = dataset.loc[i, 'bmi']
        if row >= 15 and row < 20:
            dataset.loc[i, 'bmi_range'] = '15-20'
        elif row >= 20 and row < 25:
            dataset.loc[i, 'bmi_range'] = '20-25'
        elif row >= 25 and row < 30:
            dataset.loc[i, 'bmi_range'] = '25-30'
        elif row >= 30 and row < 35:
            dataset.loc[i, 'bmi_range'] = '30-35'
        elif row >= 35 and row < 40:
            dataset.loc[i, 'bmi_range'] = '35-40'
        elif row >= 40 and row < 45:
            dataset.loc[i, 'bmi_range'] = '40-45'
        elif row >= 45 and row < 50:
            dataset.loc[i, 'bmi_range'] = '45-50'
        else:
            dataset.loc[i, 'bmi_range'] = '50+'
    return dataset
    
            
bmi_grouped_insurance_data = group_bmi_ranges(insurance_data)   
bmi_grouped_insurance_data

age     sex     bmi  children smoker     region      charges bmi_range
0      19  female  27.900         0    yes  southwest  16884.92400     25-30
1      18    male  33.770         1     no  southeast   1725.55230     30-35
2      28    male  33.000         3     no  southeast   4449.46200     30-35
3      33    male  22.705         0     no  northwest  21984.47061     20-25
4      32    male  28.880         0     no  northwest   3866.85520     25-30
...   ...     ...     ...       ...    ...        ...          ...       ...
1333   50    male  30.970         3     no  northwest  10600.54830     30-35
1334   18  female  31.920         0     no  northeast   2205.98080     30-35
1335   18  female  36.850         0     no  southeast   1629.83350     35-40
1336   21  female  25.800         0     no  southwest   2007.94500     25-30
1337   61  female  29.070         0    yes  northwest  29141.36030     25-30

[1338 rows x 8 columns]

In [39]:
charges_by_bmi_range = insurance_data[['bmi_range', 'charges', 'smoker']]
charges_by_bmi_range

bmi_range      charges smoker
0        25-30  16884.92400    yes
1        30-35   1725.55230     no
2        30-35   4449.46200     no
3        20-25  21984.47061     no
4        25-30   3866.85520     no
...        ...          ...    ...
1333     30-35  10600.54830     no
1334     30-35   2205.98080     no
1335     35-40   1629.83350     no
1336     25-30   2007.94500     no
1337     25-30  29141.36030    yes

[1338 rows x 3 columns]

16. Now I'm in a position to compare smokers against BMI ranges to see if there is a consic=stency in smokers being charged more for their insurances.

In [68]:
charges_by_bmi_range_smoker = charges_by_bmi_range.groupby(['bmi_range', 'smoker'])
charges_by_bmi_range_smoker